File: src/NexusUnleashed.Network/WorldPacket.cs
using System;
using System.Collections.Generic;
using NexusUnleashed.Cryptography;

namespace NexusUnleashed.Network;

public static class WorldPacket
{
    public const ushort ServerContainer = 0x03DC;
    public const ushort ClientContainer = 0x0244;

    public const ulong WorldChannelSeed = PacketCrypt.AuthChannelKey;

    public static byte[] EncodeServer(ushort innerOpcode, byte[] innerBody, PacketCrypt crypt)
    {
        byte[] payload = BuildContainerPayload(innerOpcode, innerBody, crypt);
        return GamePacketFrame.Encode(ServerContainer, payload);
    }

    public static byte[] EncodeClient(ushort innerOpcode, byte[] innerBody, PacketCrypt crypt)
    {
        byte[] payload = BuildContainerPayload(innerOpcode, innerBody, crypt);
        return GamePacketFrame.Encode(ClientContainer, payload);
    }

    public static (ushort Opcode, byte[] Body) DecodeContainer(byte[] containerPayload, PacketCrypt crypt)
    {
        if (containerPayload.Length < 4)
            throw new ArgumentException("container payload shorter than its length field");

        uint innerLen = ReadU32LE(containerPayload, 0);        if (innerLen < 4 || innerLen > containerPayload.Length)
            throw new ArgumentException($"container innerLen {innerLen} out of range ({containerPayload.Length})");

        int cipherLen = (int)innerLen - 4;
        var cipher = new byte[cipherLen];
        Array.Copy(containerPayload, 4, cipher, 0, cipherLen);

        byte[] inner = crypt.Decrypt(cipher, cipherLen);
        if (inner.Length < 2)
            throw new ArgumentException("decrypted inner shorter than an opcode");

        ushort opcode = (ushort)(inner[0] | (inner[1] << 8));
        var body = new byte[inner.Length - 2];
        Array.Copy(inner, 2, body, 0, body.Length);
        return (opcode, body);
    }

    private static byte[] BuildContainerPayload(ushort innerOpcode, byte[] innerBody, PacketCrypt crypt)
    {
        var inner = new byte[2 + innerBody.Length];
        inner[0] = (byte)innerOpcode;
        inner[1] = (byte)(innerOpcode >> 8);
        Array.Copy(innerBody, 0, inner, 2, innerBody.Length);

        byte[] cipher = crypt.EncryptForClient(inner, inner.Length);

        uint innerLen = (uint)(4 + cipher.Length);
        var payload = new byte[4 + cipher.Length];
        WriteU32LE(payload, 0, innerLen);
        Array.Copy(cipher, 0, payload, 4, cipher.Length);
        return payload;
    }

    private static uint ReadU32LE(byte[] b, int off)
        => (uint)(b[off] | (b[off + 1] << 8) | (b[off + 2] << 16) | (b[off + 3] << 24));

    private static void WriteU32LE(byte[] b, int off, uint v)
    {
        b[off] = (byte)v;
        b[off + 1] = (byte)(v >> 8);
        b[off + 2] = (byte)(v >> 16);
        b[off + 3] = (byte)(v >> 24);
    }
}

File: cpp/src/net/world_packet.cpp
// NexusUnleashed - clean-room authored. See world_packet.h. 1:1 with WorldPacket.cs.
#include "net/world_packet.h"
#include "net/frame.h"
#include <stdexcept>
#include <cstring>

namespace nexus::net {

static uint32_t ReadU32LE(const uint8_t* b) {
    return static_cast<uint32_t>(b[0]) | (static_cast<uint32_t>(b[1]) << 8)
         | (static_cast<uint32_t>(b[2]) << 16) | (static_cast<uint32_t>(b[3]) << 24);
}
static void WriteU32LE(uint8_t* b, uint32_t v) {
    b[0] = static_cast<uint8_t>(v); b[1] = static_cast<uint8_t>(v >> 8);
    b[2] = static_cast<uint8_t>(v >> 16); b[3] = static_cast<uint8_t>(v >> 24);
}

// container payload = [u32 innerLen self-inclusive][encrypted [u16 op][body]].
// serverToClient selects the cipher: S->C uses the QWORD cipher (EncryptForClient); C->S uses
// the BYTE cipher (Encrypt) — the two directions differ (see packet_crypt.h).
static std::vector<uint8_t> BuildContainerPayload(uint16_t innerOpcode,
                                                  const std::vector<uint8_t>& innerBody,
                                                  crypto::PacketCrypt& crypt, bool serverToClient) {
    std::vector<uint8_t> inner(2 + innerBody.size());
    inner[0] = static_cast<uint8_t>(innerOpcode);
    inner[1] = static_cast<uint8_t>(innerOpcode >> 8);
    std::memcpy(inner.data() + 2, innerBody.data(), innerBody.size());

    std::vector<uint8_t> cipher = serverToClient ? crypt.EncryptForClient(inner) : crypt.Encrypt(inner);

    std::vector<uint8_t> payload(4 + cipher.size());
    WriteU32LE(payload.data(), static_cast<uint32_t>(4 + cipher.size()));
    std::memcpy(payload.data() + 4, cipher.data(), cipher.size());
    return payload;
}

std::vector<uint8_t> WorldPacket::EncodeServer(uint16_t innerOpcode, const std::vector<uint8_t>& innerBody,
                                               crypto::PacketCrypt& crypt) {
    return GamePacketFrame::Encode(ServerContainer, BuildContainerPayload(innerOpcode, innerBody, crypt, true));
}

std::vector<uint8_t> WorldPacket::EncodeServerVia(uint16_t containerOpcode, uint16_t innerOpcode,
                                                  const std::vector<uint8_t>& innerBody, crypto::PacketCrypt& crypt) {
    return GamePacketFrame::Encode(containerOpcode, BuildContainerPayload(innerOpcode, innerBody, crypt, true));
}

std::vector<uint8_t> WorldPacket::EncodeClient(uint16_t innerOpcode, const std::vector<uint8_t>& innerBody,
                                               crypto::PacketCrypt& crypt) {
    return GamePacketFrame::Encode(ClientContainer, BuildContainerPayload(innerOpcode, innerBody, crypt, false));
}

std::pair<uint16_t, std::vector<uint8_t>> WorldPacket::DecodeContainer(const std::vector<uint8_t>& containerPayload,
                                                                       crypto::PacketCrypt& crypt) {
    if (containerPayload.size() < 4)
        throw std::invalid_argument("container payload shorter than its length field");
    uint32_t innerLen = ReadU32LE(containerPayload.data());
    if (innerLen < 4 || innerLen > containerPayload.size())
        throw std::invalid_argument("container innerLen out of range");

    size_t cipherLen = innerLen - 4;
    std::vector<uint8_t> cipher(containerPayload.begin() + 4, containerPayload.begin() + 4 + cipherLen);
    std::vector<uint8_t> inner = crypt.Decrypt(cipher);
    if (inner.size() < 2)
        throw std::invalid_argument("decrypted inner shorter than an opcode");

    uint16_t opcode = static_cast<uint16_t>(inner[0] | (inner[1] << 8));
    std::vector<uint8_t> body(inner.begin() + 2, inner.end());
    return { opcode, std::move(body) };
}

} // namespace nexus::net
